### ISSUE TITLE
cmd/link: improve error for missing method body

### ISSUE THREAD

**Issue by ianlancetaylor** [golang/go#48780]
The language permits omitting the method body.  However, the error message for this could be clearer.

Test case (uses `import "C"` to force the compiler to assume that method bodies are defined elsewhere):

```Go
package main

import "C"

type S struct{}

func (S) M()

var Sink interface{}

func main() {
	Sink = S{}
}
```

Running `go build foo.go` produces

```
type.main.S: relocation target main.S.M not defined
```

This error message is accurate but conveys little information to someone unfamiliar with linker jargon like "relocation target."  We should instead say "missing function body", though unfortunately it may be difficult to give a good file/line position for the declaration with the missing body.

**Comment by cherrymui** [comment]
Good point. I guess one difficulty is that for an undefined symbol the linker doesn't know if it is supposed to be a function or not. Maybe we could make some guess, like assuming all ABIInternal references are functions (which will probably help this one), or name-based guessing.

**Comment by cherrymui** [comment]
Or we could say "undefined function or variable".

**Comment by deepto98** [comment]
@ianlancetaylor @cherrymui Hi, I'd like to pick this issue up? Can you please assign it, if it isn't being worked on already? Also could you point me to the relevant files for the error definitions?

**Comment by cherrymui** [comment]
@deepto98 thanks. The error is emitted at https://cs.opensource.google/go/go/+/master:src/cmd/link/internal/ld/errors.go;l=63
Feel free to send a CL when you ready. (But note that I'm still unsure what the best solution is.)

**Comment by deepto98** [comment]
@cherrymui @ianlancetaylor I'm trying to reproduce this issue, but I get a different error for the code snippet given:
![image](https://user-images.githubusercontent.com/91651033/136091701-7f2c6e92-ba43-4925-bafa-85ccbc0c0de8.png)

**Comment by deepto98** [comment]
Okay, sure, will keep this in mind about error messages. I'm also on linux (Ubuntu 20.04) and the go version I'm using is **go1.12.2 gccgo (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0 linux/amd64** . I added the code to a file (foo.go) and ran **go build foo.go**

**Comment by deepto98** [comment]
@ianlancetaylor I wasn't aware about external and internal linking, and just read up on it. Do you know how I can change the config to switch to internal linking?

**Comment by cherrymui** [comment]
This issue is not about gccgo. This is about the Go toolchain distributed from https://golang.org/dl/ , also known as "the gc toolchain".

**Comment by deepto98** [comment]
@cherrymui Should I change the error message format to say **type.main.S: undefined function or variable**?

**Comment by cherrymui** [comment]
I don't know. I think it is important to print the undefined symbol name, i.e. the `main.S.M` in the original post. But I don't know the exact wording.

**Comment by deepto98** [comment]
okay, I'm looking into this

**Comment by deepto98** [comment]
@cherrymui  @ianlancetaylor from what I found out for empty variable definitions, they're already handled in **/go/src/go/parser/parser.go** with messages **missing variable type or initialization** or **missing constant value**, so maybe we can assume that ABI reference symbols are functions, as @cherrymui  had suggested?

**Comment by ianlancetaylor** [comment]
We may have to consider references from assembly code.
